.github/workflows/main-docker-build.yml

```
@@ -0,0 +1,50 @@
+name: "main-docker-build"
+
+on:
+  push:
+    branches:
+      - "main"
+    tags:
+      - "v*"
+  workflow_dispatch:
+
+env:
+  GO_VERSION: 1.19
+  IMAGE_NAME: ${{ github.repository }}
+  REGISTRY: ghcr.io
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}
+  cancel-in-progress: true
+
+jobs:
+  docker-build:
+    runs-on: "ubuntu-latest"
+    permissions:
+      contents: write
+      packages: write
+    steps:
+      - uses: "actions/checkout@v3"
+      - name: Docker meta
+        id: meta
+        uses: docker/metadata-action@v4
+        with:
+          images: ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}
+          tags: |
+            type=sha
+            type=semver,pattern={{version}}
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+      - name: Login to GHCR
+        uses: docker/login-action@v2
+        with:
+          registry: ghcr.io
+          username: ${{ github.repository_owner }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+      - name: Build and push
+        uses: docker/build-push-action@v3
+        with:
+          platforms: linux/arm64, linux/amd64
+          push: true
+          tags: ${{ steps.meta.outputs.tags }}
+          file: docker/Dockerfile
```

We can use an `if` condition to handle the different triggers. This allows us to only have the one workflow file that handles all cases. 

```suggestion
      - name: Build and Push amd Image
        if: ${{ github.event_name == 'pull_request' }}
        uses: docker/build-push-action@v3
        with:
          platforms: linux/amd64
          push: true
          tags: ${{ steps.meta.outputs.tags }}
          file: docker/Dockerfile
      - name: Build and Push All Images
        if: ${{ github.event_name != 'pull_request' }}
        uses: docker/build-push-action@v3
        with:
          platforms: linux/arm64, linux/amd64
          push: true
          tags: ${{ steps.meta.outputs.tags }}
          file: docker/Dockerfile
```